### PR TITLE
Refactor the construction of sidebar settings to prep for RPC expansion

### PR DESCRIPTION
**Pull request by lyzadanger**
~~Depends on #4243~~

This PR is the second refactor step to prepare for implementation of https://github.com/hypothesis/lms/issues/3647 on the client side.

It renames and restructures some of the functions and logic for building a sidebar settings object to prepare for expanding some of the configuration and settings for RPC functionality. The motivation is to make this code more accessible to future devs and to reduce complexity/fragility before extending it. It also improves typing.

As a result of tightening typing on the `groups` property on the `Service` configuration object, I had to do a little dance with typing in the `groups` service. While I do understand what is going on here (i.e. why I ran into typing errors), I'm not convinced I handled it ideally, so have included the change as a [separate commit](https://github.com/hypothesis/client/commit/f7020e1068a7f66fc9cbda15495fc822b877af27) for easier review.

### CODE REVIEW

File: src/sidebar/config/build-settings.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+ */
+
+/**
+ * Returns the global embedder ancestor frame.
+ *
+ * @param {number} levels - Number of ancestors levels to ascend.
+ * @param {Window=} window_
+ * @return {Window}
+ */
+function getAncestorFrame(levels, window_ = window) {
```
Comment (lyzadanger) [inline src/sidebar/config/build-settings.js:18]: No change to this function.

File: src/sidebar/config/build-settings.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+}
+
+/**
+ * Retrieve host configuration and list of groups from `targetFrame` via RPC
+ *
+ * @param {Window} targetFrame
+ * @param {string} origin - Origin filter for `window.postMessage` call
+ * @return {Promise<ConfigFromHost>}
+ */
+async function fetchConfigRPC(targetFrame, origin) {
```
Comment (lyzadanger) [inline src/sidebar/config/build-settings.js:38]: This function now only returns a `ConfigFromHost` — before it also merged it with `ConfigFromSidebar`, which is now part of `buildSettings`' remit.

File: src/sidebar/config/build-settings.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    targetFrame,
+    origin,
+    'requestConfig',
+    [],
+    3000
+  );
+  // Closure for the RPC call to scope parentFrame and origin variables.
+  const rpcCall = (method, args = [], timeout = 3000) =>
+    postMessageJsonRpc.call(targetFrame, origin, method, args, timeout);
+  return fetchGroupsAsync(configFromHost, rpcCall);
```
Comment (lyzadanger) [inline src/sidebar/config/build-settings.js:49]: Next set of changes should reduce this complexity: I plan on persisting RPC settings (e.g. reference to target frame, origins) on the resulting settings object, so this closure shouldn't be necessary here.

File: src/sidebar/config/build-settings.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+ * fill in the `groups` value(s) later when its ready. This helps speed
+ * up the loading process.
+ *
+ * @param {ConfigFromHost} configFromHost - Host configuration object. This
+ *   may be mutated with a list of groups fetched by RPC.
+ * @param {function} rpcCall - RPC method
+ *  (method, args, timeout) => Promise
+ * @return {Promise<ConfigFromHost>} - The mutated host configuration
+ */
+async function fetchGroupsAsync(configFromHost, rpcCall) {
```
Comment (lyzadanger) [inline src/sidebar/config/build-settings.js:67]: The logic for this is unchanged, but the comments and typing are (mostly) new.

File: src/sidebar/config/build-settings.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        service.groups = rpcCall('requestGroups', [index], null).catch(() => {
+          throw new Error('Unable to fetch groups');
+        });
+      }
+    });
+  }
+  return configFromHost;
+}
+
+/**
```
Comment (lyzadanger) [inline src/sidebar/config/build-settings.js:82]: Everything here and below can be considered "new construction": rework of the previous `fetchConfig` function.

File: src/sidebar/config/test/build-settings-test.js [hunk 6]
```diff
@@ -149,18 +149,20 @@ describe('sidebar/config/fetch-config', () => {
...
         const appConfig = {
-          services: [{ groups: '$rpc:requestGroups' }],
           appType: 'via',
         };
-        fakeJsonRpc.call.onFirstCall().resolves({ foo: 'baz' }); // host config
+        fakeJsonRpc.call.onFirstCall().resolves({
+          foo: 'baz',
+          services: [{ groups: '$rpc:requestGroups' }],
+        }); // host config
         fakeJsonRpc.call.onSecondCall().resolves(['group1', 'group2']); // requestGroups
```
Comment (lyzadanger) [inline src/sidebar/config/test/build-settings-test.js:164]: This test change and the change to the subsequent test is explained in this comment: https://github.com/hypothesis/client/pull/4242#discussion_r811381077

**Review by lyzadanger — COMMENTED** [review]
I regret that the diff isn't so hot because I edited and renamed the module. I've tried to add some helpful comments.

File: src/sidebar/config/build-settings.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+ * @typedef {import('../../types/config').ConfigFromSidebar} ConfigFromSidebar
+ * @typedef {import('../../types/config').ConfigFromHost} ConfigFromHost
+ * @typedef {import('../../types/config').SidebarSettings} SidebarSettings
+ */
+
+/**
+ * Returns the global embedder ancestor frame.
+ *
+ * @param {number} levels - Number of ancestors levels to ascend.
```
Comment (robertknight) [inline src/sidebar/config/build-settings.js:14]:
```suggestion
 * @param {number} levels - Number of ancestor levels to ascend.
```

**Review by robertknight — COMMENTED** [review]
I wrote these comments several days ago but realized I failed to actually submit them. Sorry about that!

**Comment by lyzadanger** [comment]
Converted to a draft for now because this is evolving a bit in my local branch and I don't want to distract focus from other priorities.

**Comment by lyzadanger** [comment]
Work continues locally on this branch and it is indeed meaningful work toward https://github.com/hypothesis/lms/issues/3647, but I don't want this PR taking up contextual brainspace right at the moment. I will close this PR and open a new one when it's time to get focus back on this.

**Pull request closed without merging**